File: 61_Math_Dice/pascal/mathdice.pas
(*
 * Ported from mathdice.bas to Pascal by krt@krt.com.au
 *
 * Compile with Free Pascal (https://www.freepascal.org/) ~
 *    fpc mathdice.pas
 *)

program MathDice;


procedure printDice( face_value: integer );
(* Prints a box with spots representing a die face for the user *)
begin
    writeln( ' ----- ' );

    if ( face_value = 1 ) then
        writeln( 'I     I' )
    else if ( ( face_value = 2 ) or ( face_value = 3 ) ) then
        writeln( 'I *   I' )
    else
        writeln( 'I * * I' );

    if ( ( face_value = 2 ) or ( face_value = 4 ) ) then
        writeln( 'I     I' )
    else if ( face_value = 6 ) then
        writeln( 'I * * I' )
    else
        writeln( 'I  *  I' );

    if ( face_value = 1 ) then
        writeln( 'I     I' )
    else if ( ( face_value = 2 ) or ( face_value = 3 ) ) then
        writeln( 'I   * I' )
    else
        writeln( 'I * * I' );

    writeln( ' ----- ' );
end;


procedure writeAtColumn( width: integer; words: string );
(* Prints <width> worth of spaces before the <words> to justify the text *)
var
    i: integer;
begin
    for i := 1 to width do
        write( ' ' );
    writeln( words );
end;


function inputNumber(): integer;
(* Get a number from the player with error checking.
   If they type a non-number, ask them again *)
var
    player_input:  string;    (* The string entered by the player *)
    player_answer: integer;   (* The converted value of the text *)
    input_error:   integer;   (* The letter's column that caused an error *)
begin

    input_error := 1;
    while ( input_error <> 0 ) do
    begin
        readln( player_input );

        val( player_input, player_answer, input_error );

        if ( input_error <> 0 ) then
            write( 'Please input a number: ' );
    end;

    inputNumber := player_answer;
end;



var
    dice1:         integer;   (* die 1 face value *)
    dice2:         integer;   (* die 2 face value *)
    answer:        integer;   (* the sum of the dice *)
    player_answer: integer;   (* The value entered by the player *)
begin
    writeAtColumn( 31, 'MATH DICE' );
    writeAtColumn( 15, 'CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY' );
    writeAtColumn( 15, '(Ported to Pascal Oct 2012 krt@krt.com.au)' );
    writeln( '' );
    writeln( '' );
    writeln( '' );

    writeln( 'THIS PROGRAM GENERATES SUCCESSIVE PICTURES OF TWO DICE.' );
    writeln( 'WHEN TWO DICE AND AN EQUAL SIGN FOLLOWED BY A QUESTION' );
    writeln( 'MARK HAVE BEEN PRINTED, TYPE YOUR ANSWER AND THE RETURN KEY.' );
    writeln( 'TO CONCLUDE THE LESSON, TYPE CONTROL-C AS YOUR ANSWER.' );
    writeln( '' );
    writeln( '' );

    while ( true ) do
    begin
        dice1 := Random( 6 ) + 1;   (* Random number between 1 and 6 (including)  *)
        dice2 := Random( 6 ) + 1;   (* Random number between 1 and 6 (including)  *)
        answer := dice1 + dice2;

        (* Show the player two dice faces *)
        printDice( dice1 );
        writeln( '   +' );
        printDice( dice2 );

        write( '      = ' );

        player_answer := inputNumber();

        if ( player_answer <> answer ) then
        begin
            (* Give the player a second chance at the answer... *)
            writeln( 'NO, COUNT THE SPOTS AND GIVE ANOTHER ANSWER.' );
            write( '      = ' );
            player_answer := inputNumber();
        end;

        if ( player_answer <> answer ) then
            writeln( 'NO, THE ANSWER IS ', answer )
        else
            writeln( 'RIGHT!' );

        writeln( '' );
        writeln( 'THE DICE ROLL AGAIN...' );
    end;
end.

File: 61_Math_Dice/javascript/mathdice.js
// MATH DICE
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;

    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");

                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(31) + "MATH DICE\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("THIS PROGRAM GENERATES SUCCESSIVE PICTURES OF TWO DICE.\n");
    print("WHEN TWO DICE AND AN EQUAL SIGN FOLLOWED BY A QUESTION\n");
    print("MARK HAVE BEEN PRINTED, TYPE YOUR ANSWER AND THE RETURN KEY.\n"),
    print("TO CONCLUDE THE LESSON, TYPE ZERO AS YOUR ANSWER.\n");
    print("\n");
    print("\n");
    n = 0;
    while (1) {
        n++;
        d = Math.floor(6 * Math.random() + 1);
        print(" ----- \n");
        if (d == 1)
            print("I     I\n");
        else if (d == 2 || d == 3)
            print("I *   I\n");
        else
            print("I * * I\n");
        if (d == 2 || d == 4)
            print("I     I\n");
        else if (d == 6)
            print("I * * I\n");
        else
            print("I  *  I\n");
        if (d == 1)
            print("I     I\n");
        else if (d == 2 || d == 3)
            print("I   * I\n");
        else
            print("I * * I\n");
        print(" ----- \n");
        print("\n");
        if (n != 2) {
            print("   +\n");
            print("\n");
            a = d;
            continue;
        }
        t = d + a;
        print("      =");
        t1 = parseInt(await input());
        if (t1 == 0)
            break;
        if (t1 != t) {
            print("NO, COUNT THE SPOTS AND GIVE ANOTHER ANSWER.\n");
            print("      =");
            t1 = parseInt(await input());
            if (t1 != t) {
                print("NO, THE ANSWER IS " + t + "\n");
            }
        }
        if (t1 == t) {
            print("RIGHT!\n");
        }
        print("\n");
        print("THE DICE ROLL AGAIN...\n");
        print("\n");
        n = 0;
    }
}

main();
